migrator/migrator.go

```diff
@@ -0,0 +1,269 @@
+	// nodeCount can only be 2 or 3
+	// 2  when adding second node to a single node etcd cluster
+	// 3 when adding third node to two node etcd cluster
+	if nodeCount != 2 && nodeCount != 3 {
+		return microerror.Maskf(executionFailedError, "nodeCount can only have values 2 or 3")
+	}
+
+	// execute commands on the node so the node in order to configure new etcd3 member to join existing cluster
+	{
+		nodeName := nodeNames[nodeCount-1]
+
+		// the final sed command may look like this:
+		// sed -i 's/--initial-cluster .*\\/--initial-cluster etcd1=https://etcd1.clusterd.domain.io:2380,etcd1=https://etcd2.clusterd.domain.io:2380/g' /etc/systemd/system/etcd3.service'
+		sedReplaceRegEx := "--initial-cluster .*\\\\"
+		sedReplaceWith := fmt.Sprintf("--initial-cluster %s\\\\", initialCluster(m.etcdStartingIndex, m.baseDomain, nodeCount))
+		sedInitialClusterCmd := fmt.Sprintf("sed -i 's/%s/%s/g' /etc/systemd/system/etcd3.service", sedReplaceRegEx, sedReplaceWith)
+
+		commands := []string{
+			"systemctl stop etcd3",          // stop etcd3 service
+			"rm -rf /var/lib/etcd/member",   // ensure the data folder is empty
+			sedInitialClusterCmd,            // sed command to properly set initialCluster string
+			"systemctl daemon-reload",       // load new etcd3 service file
+			"systemctl start etcd3.service", // restart etcd3, after this etcd3 will start syncing data from the cluster
+		}
+
+		fmt.Printf("Configuring node %s for etcd cluster.\n", nodeName)
+		// execute commands above on the node via k8s job
+		err := m.runCommandsOnNode(nodeName, commands)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	nodeIndex := m.etcdStartingIndex + nodeCount - 1
+	// add the new node to the etcd cluster via etcd client API
+	{
+		peerUrls := []string{etcdPeerName(nodeIndex, m.baseDomain)}
+		r, err := m.etcdClient.Cluster.MemberAdd(ctx, peerUrls)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		fmt.Printf("Added new member %s to the etcd cluster.\n", r.Member.PeerURLs)
+	}
+
+	// wait until k8s api is available again, as etcd data sync will make API unavailable for short time
+	err := waitForApiAvailable(m.k8sClient)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	fmt.Printf("Etcd cluster synced, node %s succesfully joined etcd cluster.\n", nodeNames[nodeIndex])
+
+	return nil
+}
+
+func getMasterNodes(c kubernetes.Interface, labelSelector string) ([]string, error) {
+	var nodeNames []string
+
+	b := backoff.NewMaxRetries(maxRetriesNodes, masterNodeFetchInterval)
+	o := func() error {
+		nodeList, err := c.CoreV1().Nodes().List(metav1.ListOptions{LabelSelector: labelSelector})
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		if len(nodeList.Items) == masterNodeCount {
+			nodeNames = getNodeNames(nodeList.Items)
+			fmt.Printf("Found %d masters %s, %s, %s.\n", masterNodeCount, nodeNames[0], nodeNames[1], nodeNames[2])
```

Comment:
Design wise, the code makes pretty strong assumptions about 3 masters already so I am not sure which value `masterNodeCount` serves. Just saying.

Reply:
This is necessery to ensure there are excatly 3 node sin k8s api. The rest of code depends on that so this retried until there are 3 nodes.